ksqldb-engine/src/main/java/io/confluent/ksql/internal/StorageUtilizationMetrics.java

```diff
@@ -0,0 +1,280 @@
+  @Override
+  public void metricChange(final KafkaMetric metric) {
+    if (!metric.metricName().name().equals("total-sst-files-size")) {
+      return;
+    }
+
+    handleNewSstFilesSizeMetric(
+      metric,
+      metric.metricName().tags().getOrDefault("task-id", ""),
+      getQueryId(metric)
+    );
+  }
+
+  @Override
+  public void metricRemoval(final KafkaMetric metric) {
+    final MetricName metricName = metric.metricName();
+    if (!metricName.name().equals("total-sst-files-size")) {
+      return;
+    }
+
+    final String queryId = getQueryId(metric);
+    final String taskId = metric.metricName().tags().getOrDefault("task-id", "");
+    final TaskStorageMetric taskMetric = metricsSeen.get(queryId).get(taskId);
+
+    handleRemovedSstFileSizeMetric(taskMetric, metric, queryId, taskId);
+  }
+
+  @Override
+  public void close() {
+  }
+
+  @Override
+  public Set<String> reconfigurableConfigs() {
+    return null;
+  }
+
+  @Override
+  public void validateReconfiguration(final Map<String, ?> configs) throws ConfigException {
+
+  }
+
+  @Override
+  public void reconfigure(final Map<String, ?> configs) {
+
+  }
+
+  @Override
+  public void contextChange(final MetricsContext metricsContext) {
+
+  }
+
+  private synchronized void handleNewSstFilesSizeMetric(
+    final KafkaMetric metric,
+    final String taskId,
+    final String queryId
+  ) {
+    // if we haven't seen a task for this query yet
+    if (!metricsSeen.containsKey(queryId)) {
+      metricRegistry.addMetric(
+          metricRegistry.metricName(
+          "query-storage-usage",
+          METRIC_GROUP,
+          ImmutableMap.of("query-id", queryId)),
+          (Gauge<BigInteger>) (config, now) -> computeQueryMetric(queryId)
+      );
+      metricsSeen.put(queryId, new HashMap<>());
+    }
+    final TaskStorageMetric newMetric;
+    // We haven't seen a metric for this query's task before
+    if (!metricsSeen.get(queryId).containsKey(taskId)) {
+      // create a new task level metric to track state store storage usage
+      newMetric = new TaskStorageMetric(
+        metricRegistry.metricName(
+          "task-storage-usage",
+          METRIC_GROUP,
+          ImmutableMap.of("task-id", taskId, "query-id", queryId)
+        ));
+      // add to list of seen task metrics for this query
+      metricsSeen.get(queryId).put(taskId, newMetric);
+      // create gauge for task level storage usage
+      metricRegistry.addMetric(
+          newMetric.metricName,
+          (Gauge<BigInteger>) (config, now) -> newMetric.getValue()
+      );
+    } else {
+      // We have this metric already
+      newMetric = metricsSeen.get(queryId).get(taskId);
+    }
+    newMetric.add(metric);
+  }
+
+  private synchronized void handleRemovedSstFileSizeMetric(
+      final TaskStorageMetric taskMetric,
+      final KafkaMetric metric,
+      final String queryId,
+      final String taskId
+  ) {
+    // remove storage metric for this task
+    taskMetric.remove(metric);
```

Comment:
See my other comment: if we would only ever have one metric under that task, then we can simplify this a bit.